cdap-api/src/main/java/co/cask/cdap/api/dataset/lib/DynamicPartitioner.java

```diff
@@ -27,6 +28,27 @@
  */
 public abstract class DynamicPartitioner<K, V> {
 
+  /**
+   * Defines options for when writing to a partition.
+   */
+  public enum PartitionWriteOption {
+    /**
+     * Will only allow writing to a new partition. If an attempt is made to write to a previously-existing partition,
+     * a {@link DataSetException} will be thrown.
+     */
+    CREATE,
+    /**
+     * Allows appending to an existing partition. If the partition does not already exist, it will be created.
+     * Existing partition's metadata is merged with the appending partition's metadata, with the appending partition's
+     * metadata taking precedence if keys conflict.
+     */
+    ALLOW_APPEND,
```

Comment:
I prefer CREATE_OR_APPEND and CREATE_OR_OVERWRITE, since the ordering indicates that first it attempts to create, and only falls back to append/overwrite, if the partition already exists.